File: Grind75_1/week1/ValidParentheses.py
"""
https://leetcode.com/problems/valid-parentheses/description/
"""

import collections

class Solution:

    def isValid(self, s: str) -> bool:
        # s is to only give parenthesis characters

        # dictionary for valid pairs of parenthesis
        dictParenPairs = {'(':')', '[':']', '{':'}'}
        # linked list as a stack to keep track of order of parenthesis. LIFO
        linkedList = collections.deque()
        
        # Iterate the characters in string 's'
        for c in s:
            if c in dictParenPairs:
                # append open parenthesis to the stack since no restriction when they can appear.
                linkedList.append(c)
            elif len(linkedList) == 0 or dictParenPairs[linkedList.pop()] != c:
                # if linked list length is 0 and already has a closing parenthesis, return false
                # or
                # if the complement of the open parenthesis on the top of the stack does not match the current character, return false
                return False
            else:
                pass

        if len(linkedList) == 0:
            # if the length of the stack is 0, it indicates all parenthesis were correctly closed
            return True
        else:
            return False

    def isValidLol(self, s: str) -> bool:

        # two dicts so if string 's' contains characters other than parenthesis
        dictParenPairs = {'(':')', '[':']', '{':'}'}
        dictParenPairsRev = {')':'(', ']':'[', '}':'{'}

        linkedList = collections.deque()
        for c in s:
            if c in dictParenPairs:
                linkedList.append(c)
            elif c in dictParenPairsRev:
                if len(linkedList) == 0:
                    return False
                else:
                    try:
                        if linkedList.pop() != dictParenPairsRev[c]:
                            # wrong pair
                            return False
                        else:
                            pass
                    except:
                        return False
            else:
                pass

        if len(linkedList) == 0:
            return True
        else:
            return False

File: Grind75/Stack/valid-parentheses.js
// https://leetcode.com/problems/valid-parentheses/description/

/**
// create Set for the valid open parentheses characters
const open = new Set(['(', '{', '['])
// create Map for the closing parentheses characters the the valid openning character
const close = new Map([
    [')', '('],
    ['}', '{'],
    [']','[']
])

create an Array as the Stack. Since an incoming parenthesis must close the most recent open parenthesis, the stack will maintain the most recent on the top.

iterate i in String s
    if (s[i] in open)
        push onto stack
    else if (s[i] in close)
        if (stack.length === 0 || the top open bracket in the stack !== appropriate open bracket for this close bracket at s[i]) {
            return false
        }
        pop stack
    // else can ignore

return true if Stack.length === 0 else false    // only true if all open brackets have been closed.

- Time: O(n)
- Space: O(n)
 */

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    // create Set for the valid open parentheses characters
    const open = new Set(['(', '{', '['])
    // create Map for the closing parentheses characters the the valid openning character
    const close = new Map([
        [')', '('],
        ['}', '{'],
        [']','[']
    ])
    const stack = new Array()

    for (let i = 0; i < s.length; i ++) {
        if (open.has(s[i])) {
            stack.push(s[i])
        } else if (close.has(s[i])) {
            if (stack.length === 0 || stack[stack.length - 1] !== close.get(s[i])) {
                return false
            }
            stack.pop()
        }
    }

    return stack.length === 0 ? true : false
};